reposcan/pkgtree.py

```diff
@@ -0,0 +1,204 @@
+            for evr_id, epoch, ver, rel in cursor:
+                self.evrid2evr[evr_id] = (epoch, ver, rel)
+
+    def load_arch(self):
+        """Load the archid2arch dict"""
+        with self._named_cursor() as cursor:
+            cursor.execute("select id, name from arch")
+            for arch_id, name in cursor:
+                self.archid2arch[arch_id] = name
+
+    def load_repodata(self):
+        """Load the repo data"""
+        with self._named_cursor() as cursor:
+            cursor.execute("""select r.id,
+                                     cs.label,
+                                     cs.name,
+                                     r.basearch_id,
+                                     r.releasever,
+                                     r.revision
+                                from repo r
+                                join content_set cs on cs.id = r.content_set_id
+                           """)
+            for repo_id, label, name, arch_id, releasever, revision in cursor:
+                self.repodata[repo_id] = {'revision': revision,
+                                          'data': {'label': label,
+                                                   'name': name,
+                                                   'arch': self.archid2arch[arch_id],
+                                                   'releasever': releasever,
+                                                   'revision':format_datetime(revision)}
+                                         }
+
+    def load_cves(self):
+        """Load CVE data"""
+        with self._named_cursor() as cursor:
+            cursor.execute("select id, name from cve")
+            for cve_id, name in cursor:
+                self.cvename[cve_id] = name
+
+    def load_errata(self):
+        """Load the errata data"""
+        with self._named_cursor() as cursor:
+            cursor.execute("select id, name, issued from errata")
+            for errata_id, name, issued in cursor:
+                self.erratadata[errata_id] = {'issued': issued,
+                                              'data': {'name': name,
+                                                       'issued': format_datetime(issued)}
+                                             }
+
+    def associate_cves_to_errata(self):
+        """Associate CVEs to errata"""
+        with self._named_cursor() as cursor:
+            cursor.execute("select errata_id, cve_id from errata_cve")
+            for errata_id, cve_id in cursor:
+                self.erratadata[errata_id]['data'].setdefault('cve_list', []).append(self.cvename[cve_id])
+
+    def load_packages(self):
+        """Load the packages info"""
+        with self._named_cursor() as cursor:
+            cursor.execute("""select p.id, p.name_id, p.evr_id, p.arch_id
+                                from package p
+                          inner join evr on p.evr_id = evr.id
+                            order by evr.evr""")
```

Comment:
@eherget Does `order by evr.evr` ensure the epoch-version-release are sorted the same way the rpm library used in yum sorts them?

Or put another way (because I shouldn't need to care about the implementation), is the pkgtree output guaranteed to have the NEVRAs ordered the same way as the rpm library orders them?

Reason I'm asking is that I **don't** want to re-use the following code that we have to have in our current (overly complex) version tree tooling.

```

import rpm  # yum install rpmdevtools is necessary for this

def sort_nevra(unsorted_nevra_list):
    """
    Sorts the items based on the name-version-release string using the rpm library,
    the same way rpmdev-sort does it, which is the same way yum sorts name-version-release strings.
    """

    class NevrVer(object):
        # taken and modified from:
        #
        # rpmdev-sort -- sort rpms from standard input
        #
        # Copyright (c) 2010 Dan Horak
        #               2010-2013 Ville Skyttä
        #
        # This program is free software; you can redistribute it and/or modify
        # it under the terms of the GNU General Public License as published by
        # the Free Software Foundation; either version 2 of the License, or
        # (at your option) any later version.
        #
        # This program is distributed in the hope that it will be useful,
        # but WITHOUT ANY WARRANTY; without even the implied warranty of
        # MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
        # GNU General Public License for more details.
        #
        # You should have received a copy of the GNU General Public License
        # along with this program; if not, write to the Free Software
        # Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA
        ver = None
        n = None
        e = None
        v = None
        r = None

        def __init__(self, ver):
            self.ver = ver
            self.e = ""
            self.n, self.v, self.r = ver.name, ver.version, ver.release

        def __cmp__(self, other):
            if self.n > other.n:
                return 1
            elif self.n < other.n:
                return -1
            return rpm.labelCompare((self.e, self.v, self.r),
                                    (other.e, other.v, other.r))

        def __lt__(self, other):
            return self.__cmp__(other) < 0

        def __repr__(self):
            return self.ver.nvr_string

    sorted_nevra_list_tmp = []
    for ver in unsorted_nevra_list:
        arr = []
        arr.append(ver)
        arr.append(NevrVer(ver))
        sorted_nevra_list_tmp.append(arr)
    sorted_nevra_list_tmp.sort(key=lambda x: x[1])
    sorted_nevra_list = [x[0] for x in sorted_nevra_list_tmp]

    return sorted_nevra_list

```

Reply:
The `order by evr.evr` makes sure that for a given package name, the list of packages/nevras is ordered.  However, the set of package names are not necessarily guaranteed to be ordered.

As for whether the list of packages/nevras for a given package name is ordered the way the rpm library orders them, I'm not certain and will reach out to someone on my team (@grinrag ) to find out.

One thing I see by examining the code you posted is that the posted code does not take epoch into account (its set to "" and doesn't appear to get modified anywhere).  Our sorting of evr compares the epoch from the package/nevra, then the version, then the release.

Reply:
So I've learned that the way we do comparison of EVR came from the implementation found in Sat5.  We are not sure where that implementation came from.  Hopefully by understanding what our code does, you might be able to determine if its the same as rpm library?

Below is a snippet of db code that builds an array of evr_array_item (we defined) items from the components of the version and release values.  These arrays and the epoch are put into a column of type evr_t (we defined), and then when we query the database we sort on this column.

When parsing a version or release where some component has both letters and numbers in it, the array of evr_array_items looks a bit strange.  Here's a couple samples of the arrays produced by the parsing code below (rpmver_array):

1.2.3 -> [(1, NULL), (2, NULL), (3, NULL)]
d38b4796  ->  [(0, 'd'), (38, NULL), (0, 'b'), (4796, NULL)]

570c8cd3 -> [(570, NULL), (0, 'c'), (8, NULL), (0, 'cd'), (3, NULL)]

svn16832.2.16b -> [(0, 'svn'), (16832, NULL), (2, NULL), (16, NULL), (0, 'b')]

```
-- -----------------------------------------------------
-- evr type
-- represents version and release as arrays of parsed components
-- of proper type (numeric or varchar)
-- this allow us to use standard array operation for = > <
-- and most importantly sorting
-- -----------------------------------------------------
create type evr_array_item as (
        n       NUMERIC,
        s       TEXT
);

create type evr_t as (
        epoch INT,
        version evr_array_item[],
        release evr_array_item[]
);

create or replace FUNCTION rpmver_array (string1 IN VARCHAR)
RETURNS evr_array_item[] as $$
declare
    str1 VARCHAR := string1;
    digits VARCHAR(10) := '0123456789';
    lc_alpha VARCHAR(27) := 'abcdefghijklmnopqrstuvwxyz';
    uc_alpha VARCHAR(27) := 'ABCDEFGHIJKLMNOPQRSTUVWXYZ';
    alpha VARCHAR(54) := lc_alpha || uc_alpha;
    one VARCHAR;
    isnum BOOLEAN;
    ver_array evr_array_item[] := ARRAY[]::evr_array_item[];
BEGIN
    if str1 is NULL
    then
        RAISE EXCEPTION 'VALUE_ERROR.';
    end if;

    one := str1;
    <<segment_loop>>
    while one <> ''
    loop
        declare
            segm1 VARCHAR;
            segm1_n NUMERIC := 0;
        begin
            -- Throw out all non-alphanum characters
            while one <> '' and not isalphanum(one)
            loop
                one := substr(one, 2);
            end loop;
            str1 := one;
            if str1 <> '' and isdigit(str1)
            then
                str1 := ltrim(str1, digits);
                isnum := true;
            else
                str1 := ltrim(str1, alpha);
                isnum := false;
            end if;
            if str1 <> ''
            then segm1 := substr(one, 1, length(one) - length(str1));
            else segm1 := one;
            end if;

            if segm1 = '' then return ver_array; end if; /* arbitrary */
            if isnum
            then
                segm1 := ltrim(segm1, '0');
                if segm1 <> '' then segm1_n := segm1::numeric; end if;
                segm1 := NULL;
            else
            end if;
            ver_array := array_append(ver_array, (segm1_n, segm1)::evr_array_item);
            one := str1;
        end;
    end loop segment_loop;

    return ver_array;
END ;
$$ language 'plpgsql';

-- -----------------------------------------------------
-- Table vmaas.evr
-- -----------------------------------------------------
CREATE TABLE IF NOT EXISTS evr (
  id SERIAL,
  epoch TEXT NOT NULL, CHECK (NOT empty(epoch)),
  version TEXT NOT NULL, CHECK (NOT empty(version)),
  release TEXT NOT NULL, CHECK (NOT empty(release)),
  evr evr_t NOT NULL,
  UNIQUE (epoch, version, release),
  PRIMARY KEY (id)
)TABLESPACE pg_default;
```

Reply:
> One thing I see by examining the code you posted is that the posted code does not take epoch into account (its set to "" and doesn't appear to get modified anywhere). Our sorting of evr compares the epoch from the package/nevra, then the version, then the release.

That is correct and so far, intentional - the data sources we currently use don't provide the epoch information and the packages we had to write detection for in the last 3 years didn't have such a versioning that omitting Epoch would break it. Especially now that we (will) take the N(E)VR(A) ordering directly from VMaaS, the only instance where omitting Epoch on our part would be a problem is the case where (Name,Version,Release) is exactly the same string, and given RH's tendency to bump Release every time, that's probably not going to happen. But yes, if we were doing the sort of thing you do (detection for everything, not just hand-picked CSAws), we might have a problem with the code I posted.